src/details/ArborX_DetailsHalfTraversal.hpp

```diff
@@ -0,0 +1,88 @@
+/****************************************************************************
+ * Copyright (c) 2017-2022 by the ArborX authors                            *
+ * All rights reserved.                                                     *
+ *                                                                          *
+ * This file is part of the ArborX library. ArborX is                       *
+ * distributed under a BSD 3-clause license. For the licensing terms see    *
+ * the LICENSE file in the top-level directory.                             *
+ *                                                                          *
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+
+#ifndef ARBORX_DETAILS_HALF_TRAVERSAL_HPP
+#define ARBORX_DETAILS_HALF_TRAVERSAL_HPP
+
+#include <ArborX_DetailsHappyTreeFriends.hpp>
+#include <ArborX_DetailsNode.hpp> // ROPE_SENTINEL
+
+namespace ArborX::Details
+{
+
+template <class BVH, class PredicateGetter, class Callback>
+struct HalfTraversal
+{
+  BVH _bvh;
+  PredicateGetter _predicate;
+  Callback _callback;
+
+  template <class ExecutionSpace>
+  HalfTraversal(ExecutionSpace const &space, BVH const &bvh,
+                PredicateGetter const &predicate, Callback const &callback)
+      : _bvh{bvh}
+      , _predicate{predicate}
+      , _callback{callback}
+  {
+    if (_bvh.empty())
+    {
+      // do nothing
+    }
+    else if (_bvh.size() == 1)
+    {
+      // do nothing either
+    }
+    else
+    {
+      auto const n = _bvh.size();
+      Kokkos::parallel_for(
+          "ArborX::HalfTraversal",
+          Kokkos::RangePolicy<ExecutionSpace>(space, n - 1, 2 * n - 1), *this);
+    }
+  }
+
+  KOKKOS_FUNCTION void operator()(int i) const
+  {
+    auto const predicate =
+        _predicate(HappyTreeFriends::getBoundingVolume(_bvh, i));
+    auto const leaf_permutation_i =
+        HappyTreeFriends::getLeafPermutationIndex(_bvh, i);
+
+    int node;
+    int next = HappyTreeFriends::getRope(_bvh, i);
+    while (next != ROPE_SENTINEL)
+    {
+      node = next;
+
+      if (predicate(HappyTreeFriends::getBoundingVolume(_bvh, node)))
+      {
+        if (!HappyTreeFriends::isLeaf(_bvh, node))
+        {
+          next = HappyTreeFriends::getLeftChild(_bvh, node);
+        }
+        else
+        {
+          _callback(leaf_permutation_i,
+                    HappyTreeFriends::getLeafPermutationIndex(_bvh, node));
```

Comment:
This callback is different from our usual as it uses `(int, int)` and not `(query, int)`. I think I understand that you don't want to attach indices to queries. At the same time, I think it may be too restrictive, as it does lose some information, for example the used radius or something similar. While I'm not going to block this, my preference would be to switch to using queries with attachment unless those have heavy penalty.

This is good, no early exit. Should be add a `static_assert` somewhere to make sure that the `_callback` is returning void?